edb/edgeql/compiler/viewgen.py

```diff
@@ -446,6 +473,133 @@ def gen_pointers_from_defaults(
+    ctx: context.ContextLevel,
+) -> List[Tuple[s_pointers.Pointer, irast.Set | None]]:
+    result: List[Tuple[s_pointers.Pointer, irast.Set | None]] = []
+
+    path_id = ir_set.path_id
+    scls_pointers = stype.get_pointers(ctx.env.schema)
+
+    # init set that will be used for __subject__
+    subject_path = irast.PathId.from_type(
+        ctx.env.schema,
+        stype,
+        typename=sn.QualName(module="__derived__", name="__subject__"),
+    )
+    subject_set = irast.Set(
+        path_id=subject_path,
+        typeref=subject_path.target,
+        shape=tuple(
+            [
+                (ptr_set, qlast.ShapeOp.ASSIGN)
+                for ptr_set in pointers.values()
+                if ptr_set
+            ]
+        ),
+    )
+    ctx.env.set_types[subject_set] = stype
+
+    # init reference to std::bool
+    bool_type = ctx.env.schema.get("std::bool", type=s_types.Type)
+    bool_path = irast.PathId.from_type(
+        ctx.env.schema,
+        bool_type,
+        typename=sn.QualName(module="std", name="bool"),
+    )
+
+    # init set that will be used for __specified__
+    specified_path = irast.PathId.from_type(
+        ctx.env.schema,
+        stype,  # TODO: what's the type of this? named tuple?
+        typename=sn.QualName(module="__derived__", name="__specified__"),
+    )
+    specified_set = irast.Set(
+        path_id=specified_path,
+        typeref=specified_path.target,
+        expr=irast.Tuple(
+            named=True,
+            typeref=specified_path.target,
+            elements=[
+                irast.TupleElement(
+                    name=pn.name,
+                    val=irast.Set(
+                        path_id=bool_path,
+                        typeref=bool_path.target,
+                        expr=irast.BooleanConstant(
+                            value="true" if pn in specified_ptrs else "false",
+                            typeref=bool_path.target,
+                        ),
+                    ),
+                )
+                for pn, _ in scls_pointers.items(ctx.env.schema)
+            ],
+        ),
+    )
+    ctx.env.set_types[specified_set] = stype
+
+    for _, ptrcls in scls_pointers.items(ctx.env.schema):
+        rewrite = ptrcls.get_rewrite(ctx.env.schema, rewrite_kind)
+        if not rewrite:
+            continue
+
+        rewrite_expr = rewrite.get_expr(ctx.env.schema)
+        assert rewrite_expr
+
+        # if pn in specified_ptrs or ptrcls.is_pure_computable(ctx.env.schema):
+        #     continue
+
+        with ctx.new() as scopectx:
+            scopectx.anchors["__specified__"] = specified_set
+            scopectx.anchors["__subject__"] = subject_set
+            scopectx.partial_path_prefix = subject_set
+
+            # scopectx.env.path_scope.attach_path(subject_set.path_id, context=None)
+            scopectx.env.singletons.append(subject_set.path_id)
+            # scopectx.iterator_path_ids |= {subject_set.path_id}
```

Comment:
My rudimentary way of setting a singleton during compilation :D

Reply:
Hm... the `attach_path` approach should be the right one